tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -163,6 +168,7 @@ void InvocationOptions::profile(llvm::FoldingSetNodeID &ID) const {
   for (auto &Arg : Args)
     ID.AddString(Arg);
   ID.AddString(PrimaryFile);
+  ID.AddPointer(FileSystem.get());
```

Comment:
I have started to think that the FileSystem pointer should not actually be part of this cache key. There is already other logic that checks the file contents and triggers recompilation when any files have changed. So including the FileSystem pointer in the cache key just causes unnecessary recompilations when we pass in new FileSystems that don't change any relevant files.

Reply:
Makes sense.

Reply:
Actually, having the VFS inside the invocation options is not going to work if we don't profile it. You'll never reliably be able to change the VFS, because an existing ASTProducer for the same file could still be in the cache.  Even closing the document and re-opening does not guarantee the cached value won't be used unless you use the special flag `key.removecache: 1`.

I'm going to pull it out of the invocation options entirely and thread it through elsewhere.  The semantics will be that a request can provide a VFS, in which case we use that VFS during `shouldRebuild`, and if the AST needs to be rebuilt, we will use that VFS for the build. Otherwise, we will fallback to getting the VFS from the editor document of the primary file.